cvmfs/catalog_balancer_impl.h

```diff
@@ -0,0 +1,189 @@
+
+  DirectoryEntryBase deb;
+  deb.name_ = NameString(name);
+  deb.mode_ = S_IFREG | S_IRWXU;
+  deb.checksum_ = file_hash;
+  deb.mtime_ = time(NULL);
+  deb.uid_ = uid;
+  deb.gid_ = gid;
+  return deb;
+}
+
+template <class CatalogMgrT>
+void CatalogBalancer<CatalogMgrT>::AddCatalogMarker(string path) {
+  XattrList xattr;
+  DirectoryEntry parent;
+  bool retval;
+  retval = catalog_mgr_->LookupPath(PathString(path), kLookupSole, &parent);
+  assert(retval);
+  DirectoryEntryBase cvmfscatalog =
+      MakeEmptyDirectoryEntryBase(".cvmfscatalog", parent.uid(),
+                                           parent.gid());
+  DirectoryEntryBase cvmfsautocatalog =
+      MakeEmptyDirectoryEntryBase(".cvmfsautocatalog", parent.uid(),
+                                           parent.gid());
+  string relative_path = path.substr(1);
+  catalog_mgr_->AddFile(cvmfscatalog, xattr, relative_path);
+  catalog_mgr_->AddFile(cvmfsautocatalog, xattr, relative_path);
+}
+
+template <class CatalogMgrT>
+void CatalogBalancer<CatalogMgrT>::Balance(catalog_t *catalog) {
+  if (catalog == NULL) {
+    // obtain a copy of the catalogs
+    vector<catalog_t*> catalogs = catalog_mgr_->GetCatalogs();
+    // we need to reverse the catalog list in order to analyze the
+    // last added ones first. This is necessary in the weird case the child
+    // catalog's underflow provokes an overflow in the father
+    reverse(catalogs.begin(), catalogs.end());
+    for (unsigned i = 0; i < catalogs.size(); ++i)
+      Balance(catalogs[i]);
+    return;
+  }
+  string catalog_path = catalog->path().ToString();
+  virtual_node_t root_node(catalog_path, catalog_mgr_);
+  root_node.ExtractChildren(catalog_mgr_);
+  // we have just recursively loaded the entire virtual tree!
+  PartitionOptimally(&root_node);
+}
+
+template <class CatalogMgrT>
+void CatalogBalancer<CatalogMgrT>::PartitionOptimally(
+    virtual_node_t *virtual_node) {
+  // postorder track of the fs-tree
+  for (unsigned i = 0; i < virtual_node->children.size(); ++i) {
+    virtual_node_t *virtual_child =
+        &virtual_node->children[i];
+    if (virtual_child->IsDirectory() && !virtual_child->IsCatalog())
+      PartitionOptimally(virtual_child);
+  }
+  virtual_node->FixWeight();
+  while (virtual_node->weight > catalog_mgr_->balance_weight_) {
+    virtual_node_t *heaviest_node =
+        MaxChild(virtual_node);
+    // we directly add a catalog in this node
+    // TODO(molina) apply heuristics here
+    if (heaviest_node != NULL &&
+        heaviest_node->weight >= catalog_mgr_->min_weight_) {
+      // the catalog now generated _cannot_ be overflowed because the tree is
+      // being traversed in postorder, handling the lightest nodes first
+      unsigned max_weight = heaviest_node->weight;
+      AddCatalogMarker(heaviest_node->path);
+      AddCatalog(heaviest_node);
+      virtual_node->weight -= (max_weight - 1);
+    } else {
+      // there is no possibility for this directory to be a catalog
```

Comment:
It would make sense to print a warning at this point.  This could be a directory with very many files, for instance, and the user might be interested to know about it.